### PR TITLE
Mechanism to close the created Gradle Check AUTOCUT flaky test issues

### PR DESCRIPTION
**Pull request by prudhvigodithi**
### Description

- Updated the logic to query the cluster for flaky tests with user defined `timeFrame` rather that default to an monthly created index. This can be updated dynamically without the library release using `timeFrame` arg.

- New library `gradleCheckFlakyTestGitHubIssue` to handle the issue creation, edit issue body and skip when issue is closed. This library is inspired from `createGithubIssue` but used only by `gradleCheckFlakyTestDetector` with some additional logic used for only gradle check flaky issue creation automation.

-  New groovy class `MarkdownComparator` to compare two markdown tables in the format created by class `CreateMarkDownTable`. This uses the row part of the markdown table and identifies if there is any difference found between the data part of the OpenSearch metrics cluster and part of the created issue body. If there is no difference the `gradleCheckFlakyTestGitHubIssue` wont re-open the issue closed by the user. If found difference the `gradleCheckFlakyTestGitHubIssue` will re-open and update the issue body with the new data.

```
Differences found:
Git Reference: f8213b8492b213922bea35d22c4317a79786a74f, Merged Pull Request: [14357](https://github.com/opensearch-project/OpenSearch/pull/14357), Build Details: [41268](https://build.ci.opensearch.org/job/gradle-check/41268/testReport/), Test Name: `org.opensearch.indices.IndicesRequestCacheIT.testCacheCleanupWithDefaultSettings {p0={"search.concurrent_segment_search.enabled":"false"}}`
```

- New class `ParseMarkDownTable` is added to just parse the markdown tables in the format created by class `CreateMarkDownTable`. The `ParseMarkDownTable` is used for parsing the issue body and for the generated test report from data from Metrics cluster.

- Added `token: ${{ secrets.CODECOV_TOKEN }}` to fix the error
```
[2024-06-25T00:24:29.694Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 429 - {'detail': ErrorDetail(string='Rate limit reached. Please upload with the Codecov repository upload token to resolve issue. Expected time to availability: 1559s.', code='throttled')}
[2024-06-25T00:24:29.695Z] ['info'] Codecov will exit with status code 0. If you are expecting a non-zero exit code, please pass in the `-Z` flag
```

### Issues Resolved
Part of https://github.com/opensearch-project/OpenSearch/issues/14475

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
The reason the `codecov/patch — 77.84% of diff hit (target 87.12%)` is failed because the `target: auto` so compares the percentage with the current head commit. Changed to `target: 70%` ( From [OpenSearch](https://github.com/opensearch-project/OpenSearch/blob/main/codecov.yml) and [build](https://github.com/opensearch-project/opensearch-build/blob/main/.codecov.yml) repo). Overall with this PR the `main` branch codeCov will be  `84.23%`.
Thank you

**Comment by rishabh6788** [comment]
Looks fine overall, didn't really get the comparison logic. 
Can you help explain how it will handle closed PRs, e.g., in https://github.com/opensearch-project/OpenSearch/issues/14472 the [PR mentioned ](https://github.com/opensearch-project/OpenSearch/pull/14494)in the table was closed and the change was merged in a different PR. 
Is this case also handled?

**Comment by prudhvigodithi** [comment]
Hey @rishabh6788  the logic assumes that the AUTOCUT issue created by the automation is fixed by the assigned owner, who then closes the issue. If the automation identifies a flaky test again (e.g., within the last 30 days and after the fix), and the issue is closed, the automation will compare the issue body table for all tests against the actual test results from the cluster query. If they match, the issue remains closed. However, if the query results differ (e.g., different post-merge commit or test results), the automation will reopen the issue and update the issue body with the new failing test data and linking the pull requests. The automation wont compare the open/closed status of past or present pull requests, once the issue is fixed and the pull requests are re-based, the flaky error should no longer be present. If the flaky error persists, the user can reopen the issue, and the automation will continue to update the issue body and link the pull requests.
Thank you 
@gaiksaya @getsaurabh02

**Review by rishabh6788 — APPROVED** [review]

File: tests/gradlecheck/FetchPostMergeFailedTestClassTest.groovy [hunk 1]
```diff
@@ -88,16 +110,16 @@ class FetchPostMergeFailedTestClassTest {
...
 
         assert result == expectedOutput
     }
 
     @Test
     void testGetPostMergeFailedTestClassReturnsKeys() {
         def expectedOutput = ["testClass1", "testClass2"]
-
-        def result = fetchPostMergeFailedTestClass.getPostMergeFailedTestClass()
+        def timeFrame = "15d"
```
Comment (gaiksaya) [inline tests/gradlecheck/FetchPostMergeFailedTestClassTest.groovy:121]: Should this be taken as an arg?

File: vars/gradleCheckFlakyTestDetector.groovy [hunk 2]
```diff
@@ -28,7 +29,8 @@ void call(Map args = [:]) {
             def awsAccessKey = env.AWS_ACCESS_KEY_ID
             def awsSecretKey = env.AWS_SECRET_ACCESS_KEY
             def awsSessionToken = env.AWS_SESSION_TOKEN
```
Comment (gaiksaya) [inline vars/gradleCheckFlakyTestDetector.groovy:31]: Where are we setting these values? Should be coming from secrets right?

File: tests/gradlecheck/FetchPostMergeFailedTestClassTest.groovy (hunk 1 above)
Comment (prudhvigodithi) [reply tests/gradlecheck/FetchPostMergeFailedTestClassTest.groovy:121]: Already its part of the arg sayali using `timeFrame`, this file `FetchPostMergeFailedTestClassTest` is the test cases for generating the query.

File: vars/gradleCheckFlakyTestDetector.groovy (hunk 2 above)
Comment (prudhvigodithi) [reply vars/gradleCheckFlakyTestDetector.groovy:31]: Yes Sayali the values are coming from secret, `withAWS(role: 'OpenSearchJenkinsAccessRole', roleAccount: "${METRICS_HOST_ACCOUNT}", duration: 900, roleSessionName: 'jenkins-session')` when used `withAWS` they are by default added to the `env` and the library uses those to create the sigv4 request.

Comment (gaiksaya) [reply vars/gradleCheckFlakyTestDetector.groovy:31]: I see, is there are possibility of exposing those secrets? I know, jenkins take care of everything that is within `withCredentials` block, but these env variables are coming as a subset of that. 
Just want to confirm we are not exposing the credentials. (even though temporary) while doing curl or other activities.

Comment (prudhvigodithi) [reply vars/gradleCheckFlakyTestDetector.groovy:31]: It wont be exposed Sayali, also we use `set +x` which disables debugging mode in the shell [OpenSearchMetricsQuery.groovy](https://github.com/opensearch-project/opensearch-build-libraries/blob/main/src/gradlecheck/OpenSearchMetricsQuery.groovy)., here is an example build https://build.ci.opensearch.org/job/gradle-check-flaky-test-detector/55/console.

**Review by gaiksaya — COMMENTED** [review]
I believe this will need version bump?

**Review by gaiksaya — APPROVED** [review]

**Comment by prudhvigodithi** [comment]
> I believe this will need version bump?

Ya let me update the version Sayali.

**Comment by prudhvigodithi** [comment]
Thanks @gaiksaya and @rishabh6788 I will go ahead and merge this PR.

**Pull request merged**